go/kbfs/libkbfs/leveldb.go

```diff
@@ -118,6 +123,41 @@ func openLevelDBWithOptions(stor storage.Storage, options *opt.Options) (
 	return &LevelDb{db, stor}, nil
 }
 
+var initialStartTime = time.Now()
+
+// openLevelDBStorage opens read-write leveldb storage based on a path.
+func openLevelDBStorage(
+	ctx context.Context, p string, log logger.Logger) (storage.Storage, error) {
+	// If we can't open the storage file because another process is
+	// still alive that has it open, wait a bit for that other process
+	// to be killed and release the resource.  Count the time from the
+	// beginning of the process's initialization, so we don't wait
+	// N*timeout for every DB we try to open.
+	eagainDeadline := initialStartTime.Add(leveldbMaxWaitOnEagain)
+	logged := false
+	for time.Now().Before(eagainDeadline) {
+		ldb, err := storage.OpenFile(p, false)
+		switch errors.Cause(err) {
+		case nil:
+			return ldb, nil
+		case syscall.EAGAIN:
+			if !logged {
+				if log != nil {
+					log.CDebugf(ctx,
+						"Waiting %s for the db %s to become available: %+v",
+						leveldbMaxWaitOnEagain, p, err)
+				}
+				logged = true
+			}
+			// Sleep and try again.
+			time.Sleep(100 * time.Millisecond)
```

Comment:
Should we be blocking the main goroutine?

Reply:
I don't see a way not to.  The calling thread is what requires the caches to be initialized.  We'd have to rewrite everything to be non-blocking when opening storage layers.